subsys/bluetooth/host/l2cap.c

```diff
@@ -2596,6 +2603,7 @@ int bt_l2cap_chan_send(struct bt_l2cap_chan *chan, struct net_buf *buf)
 			return data_sent(buf)->len;
 		}
 		BT_ERR("failed to send message %d", err);
+		net_buf_unref(buf);
```

Comment:
This doesn't look right at all. What about all the other branches of this function that may return error but do not unref the buffer?

Reply:
In those cases the buffer is always queued to be send latter (whenever there are more credits, TX context, etc).

Reply:
In what cases? I don't see any queuing happening e.g for the `-EINVAL`, `-ENOTCONN` or `-ESHUTDOWN` branches.

Reply:
Actually taking a looks at the failures case it is more common to not unref than it is, since for error like -ENOTCONN and -ESHUTDOWN we don't unref the buffer, I wonder if it would be the case of never unref even if no segments could be send which is currently the only case where the buffer would be unref which perhaps we could consider as a bug as well given that was not inline with how we treat other errors.

Reply:
@jhedberg I means in case of -ENOTCONN and -ESHUTDOWN the application would have to call net_buf_unref while other (unlikely) errors it would not have to, so Id suggested to align with the former error handling so the application always have to unref in case of error which I believe was always the intention of this API.

Reply:
Always having to unref is fine with me

Reply:
Ok, so I removed the net_buf_unref from bt_l2cap_chan_send and added a note clarifying the buffer handling to make it clearer.

Reply:
@joerchan I hope you are fine with this as well.